jobsdb/jobsdb.go

```diff
@@ -1911,6 +1923,53 @@ func (jd *HandleT) GetPileUpCounts(ctx context.Context) (map[string]map[string]i
 	return statMap, nil
 }
 
+func (jd *HandleT) GetActiveWorkspaces(ctx context.Context) ([]string, error) {
+	if !jd.dsMigrationLock.RTryLockWithCtx(ctx) {
+		return nil, fmt.Errorf("could not acquire a migration read lock: %w", ctx.Err())
+	}
+	defer jd.dsMigrationLock.RUnlock()
+	if !jd.dsListLock.RTryLockWithCtx(ctx) {
+		return nil, fmt.Errorf("could not acquire a dslist read lock: %w", ctx.Err())
+	}
+	defer jd.dsListLock.RUnlock()
+	dsList := jd.getDSList()
+
+	var workspaceIds []string
+	var queries []string
+	preciseActiveWsQuery := jd.preciseActiveWsQuery
+	for _, ds := range dsList {
+		if !preciseActiveWsQuery {
+			queries = append(queries, fmt.Sprintf(`SELECT * FROM (WITH RECURSIVE t AS (
+			(SELECT workspace_id FROM %[1]q ORDER BY workspace_id LIMIT 1)
+			UNION ALL
+			(SELECT s.* FROM t, LATERAL(
+			  SELECT workspace_id FROM %[1]q f
+			  WHERE f.workspace_id > t.workspace_id
+			  ORDER BY workspace_id LIMIT 1) s)
+		  )
+		  SELECT * FROM t) a`, ds.JobTable))
+		} else {
+			queries = append(queries, fmt.Sprintf(`SELECT DISTINCT workspace_id from %[1]q j LEFT JOIN "v_last_%[2]s" js ON js.job_id = j.job_id
+			WHERE js.job_state is NULL or js.job_state not in ('aborted', 'succeeded', 'migrated')`, ds.JobTable, ds.JobStatusTable))
+		}
```

Comment:
_**Note:**_ 
To assess what is more valuable, a fast query or a precise one, taking into consideration [this](https://www.notion.so/rudderstacks/Design-Proposal-Processor-Isolation-2db79f6c5b0f4a9fb14465b62aaaf54a#8bb2901bd4374bdfaaf30d2ca823b0e3), along with the fact that jobsdb's cache is not enabled for the `gw` rightmost dataset

Reply:
What do you think about being redundant with the workspace_id and have it in the status table too and hit just that index without a join? I think I would've find it useful to have it there a few times.

Reply:
Also, what is the performance difference between the two? Is it negligible with the number of workspaces that we deal with?

Reply:
Overall performance difference is negligible when table count is low. It becomes more prominent when table count grows. With the number of workspaces we are having it should be safer to use the index-only scan, since we can tolerate false positives better than large table counts.

With respect to the status table, it is not enough on its own, since the status table doesn't hold entries for unprocessed events, which we need to be aware of.